### ISSUE TITLE
[Feature Request] Calibration graph

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs4anki#151]
**Which module is related to your feature request?**
Optimizer.

**Describe the solution you'd like**
On https://supermemo.guru/wiki/Universal_metric_for_cross-comparison_of_spaced_repetition_algorithms there is this graph:
![image](https://user-images.githubusercontent.com/83031600/218251503-69b015b2-e5e5-4136-be4b-60b57e1354f6.png)
It would be nice if the optimizer generated something like this, predicted `r` vs measured `r` (I assume he used some sort of averaging, since in reality r is either 0 or 1). If it looks mostly like a straight line - good. If it looks like this (just an example):
![image](https://user-images.githubusercontent.com/83031600/218251846-0f8d2541-ce97-472d-9cb0-89e67f1aa53a.png)
Then there is a problem and the algorithm is performing poorly when it comes to small values of r.
Basically, this can be used to identify weaknesses of FSRS.

**Comment by L-M-Sherlock** [comment]
I am curious to know how Woz measured the retrievability. How much data is needed to draw that graph? I'm afraid the measured retrievability is inaccurate if the data is small.

**Comment by L-M-Sherlock** [comment]
I have seen a similar graph:

![image](https://user-images.githubusercontent.com/32575846/218252297-8dbd83ef-d35a-40b2-a910-5604716e5046.png)

https://papousek.github.io/analysis-of-half-life-regression-model-made-by-duolingo.html

Do you think it's your need?

**Comment by Expertium** [comment]
> I am curious to know how Woz measured the retrievability. How much data is needed to draw that graph? I'm afraid the measured retrievability is inaccurate if the data is small.
That page says "A dataset with 100,000 repetitions was used to produce this graph."

**Comment by L-M-Sherlock** [comment]
> That page says "A dataset with 100,000 repetitions was used to produce this graph."

I guess most users don't have so many repetitions. Even though I have used Anki for three years with 330,000 repetitions.

**Comment by L-M-Sherlock** [comment]
I add the calibration graph in section 4.2. You can test it for yourself by clicking the link below:
https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Feat/Calibration-graph/fsrs4anki_optimizer.ipynb

Here is the calibration graph in my dataset:

![image](https://user-images.githubusercontent.com/32575846/231331889-c1d330fc-f42d-4820-9402-589c7f73f3bc.png)

**Comment by user1823** [comment]
It is not clear what the purpose of this graph is. What is the user supposed to infer from it? What is the user supposed to do after reading the graph?

Also, what do "Observation", "Prediction" and "Number of predictions" mean here?

So, I suggest adding documentation for this new feature (and possibly, also changing the three terms mentioned above to make it clearer).

**Comment by L-M-Sherlock** [comment]
> What is the user supposed to infer from it?

Ideally, the blue line should be aligned with the orange one. It means if FSRS predicts that a card's retention is 80%, the true retention is also 80%. If the blue line is higher than the orange line, the FSRS overestimates the retention

**Comment by user1823** [comment]
How does it know the true retention achieved by using FSRS? Does it have a way to determine which reviews in the revlogs were scheduled by FSRS and which were scheduled by Anki SM-2?

Also, the optimizer generates new parameters if run after doing reviews for some days. So, is the calibration based on the previous parameters or the new parameters?

And let's say that the graph shows me that FSRS is overestimating my retention, then what am I supposed to do to correct it?

Also, what does the "Number of predictions" exactly mean in this graph?

> If the blue line is higher than the orange line, the FSRS overestimates the retention.

Shouldn't it be the reverse? (Observation > Prediction)

And, if "observation' implies "actual retention" and "prediction" implies "predicted retention", I suggest that these terms should be replaced with the latter terms.

**Comment by L-M-Sherlock** [comment]
Let's explain it step by step:

1. FSRS predicts the stability for each review.
2. The predicted retention of each review could be calculated via the exponential forgetting formula.
3. Group the reviews by their predicted retention into 20 bins.
4. Count the true retention of each bin.
5. Plot (predicted retention, true retention) in the line graph.
6. Plot (predicted retention, size of bin) in the bar graph.

Now, you will see the final graph.

**Comment by Expertium** [comment]
> Shouldn't it be the reverse? (Observation > Prediction)
Yeah, I think you're right, L-M-Sherlock probably mixed it up.

>Also, what does the "Number of predictions" exactly mean in this graph?
It shows the number of reviews, for which predictions made by FSRS fell into a certain "bin", which is just a range of values. For example, a bin might be "from 0.82 to 0.84". FSRS predicted 0.83, so it falls into that bin. Note that due to the way spaced repetition is used by most people (maintaining high retention, above 0.8 usually), the number of predictions isn't going to be equally distributed across all bins, most predictions will be above 0.8 for an average user, and only a handful of predictions will be below. This makes it hard to tell if FSRS is working well for smaller values of R (low predicted R), unless you have an absolutely humongous nubmer of reviews.

> And let's say that the graph shows me that FSRS is overestimating my retention, then what am I supposed to do to correct it?
I'm not sure if there is anything the user can do to improve calibration.

> And, if "observation' implies "actual retention" and "prediction" implies "predicted retention", I suggest that these terms should be replaced with the latter terms.
I second this, it's a bit more clear.

Lol, Sherlock posted his reply right when I was finishing mine

**Comment by L-M-Sherlock** [comment]
> Shouldn't it be the reverse? (Observation > Prediction)

Sorry for my mistake. It means FSRS underestimated the retention.

**Comment by L-M-Sherlock** [comment]
> I second this, it's a bit more clear.

Do you have any other suggestions? Maybe I need add a caption for the graph.

**Comment by Expertium** [comment]
> Do you have any other suggestions?

Add R-squared (coefficient of determination), like in the very first picture I posted in this issue, the one from Supermemo wiki. This way we can quantify calibration instead of just saying "Hmmm, I think these two lines don't quite align". Also, you should probably write a much more detailed explanation of what this chart shows in section 4.2.

Now, I might be getting a bit annoying, but I think you should try my other idea - improved formula for R, with 2 new parameters, R=e^(-t/S), R'=B*R+(1-R)*A, 0 ≤ A < B ≤ 1, R' is the new output instead of R. It's easy to implement (it's like 5-10 lines of code) and easy to test, either by comparing the values of the loss function (loss with the old formula vs with the new one), or by comparing coefficients of determination (again, with the old formula vs with the new one) for calibration. It could improve calibration.

**Comment by L-M-Sherlock** [comment]
> It's easy to implement (it's like 5-10 lines of code)

It could be easy to implement it in the scheduler. But it's difficult to do that in the optimizer. By the way, if I update the formula of FSRS, all users need to re-train after the update. It could cause some confusion. So I'm cautious in here.

**Comment by L-M-Sherlock** [comment]
> Add R-squared (coefficient of determination), like in the very first picture I posted in this issue, the one from Supermemo wiki.

It's a little weird. Woz fitted the graph with a linear function. The R-squared is used to evaluate the linear function instead of the prediction.

**Comment by Expertium** [comment]
> > Add R-squared (coefficient of determination), like in the very first picture I posted in this issue, the one from Supermemo wiki.
> 
> It's a little weird. Woz fitted the graph with a linear function. The R-squared is used to evaluate the linear function instead of the prediction.

Is that so? I thought he compared predicted R and measured R. Anyway, I recommend calculating R-squared between what you call "Average observation" and "Optimal average observation".

**Comment by Expertium** [comment]
Btw, calibration doesn't look great on my collection. I've tried it with separate decks and got similar results
![Calibration graph (Entire collection)](https://user-images.githubusercontent.com/83031600/231520566-8039c37c-599d-458d-b933-5c8f5907a4ca.png)

**Comment by L-M-Sherlock** [comment]
<img width="694" alt="image" src="https://user-images.githubusercontent.com/32575846/231642259-a720114d-a24a-4eea-8bdb-7ebc3a41b014.png">

**Comment by L-M-Sherlock** [comment]
> Btw, calibration doesn't look great on my collection. I've tried it with separate decks and got similar results

I will improve it in the next version.

**Comment by Expertium** [comment]
How are you planning to improve it?

**Comment by L-M-Sherlock** [comment]
> How are you planning to improve it?

In my recent research (unpublished), the error of power functions is lower than that of exponential functions in fitting the forgetting curve. I will test it in some users' collections.

**Comment by Expertium** [comment]
Interesting. Woz says that the forgetting curve is definitely exponential, but when you have both difficult and easy material in the same collection, the resulting curve (a superposition of different exponential curves) is best approximated with a power function.

https://web.archive.org/web/20221127081631/https://supermemo.guru/wiki/Exponential_nature_of_forgetting#Power_law_emerges_in_superposition_of_exponential_forgetting_curves

**Comment by L-M-Sherlock** [comment]
> Interesting. Woz says that the forgetting curve is definitely exponential, but when you have both difficult and easy material in the same collection, the resulting curve (a superposition of different exponential curves) is best approximated with a power function.

Yep, I tested the power function in your shared collection, and the loss decreased by 0.02. But its benefit to the scheduler is still unclear.

**Comment by L-M-Sherlock** [comment]
But it is off-topic from the current issue. We can discuss it in a new issue. Here is the latest optimizer: https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Feat/Calibration-graph/fsrs4anki_optimizer.ipynb

I will release it if there is not any problem.

**Comment by Expertium** [comment]
I recommend writing more a more detailed explanation in 4.2. Right now it just says "Ideally, the blue line should be aligned with the orange one."
